### ISSUE TITLE
Remove @RestrictTo from FirebaseAdapter

### ISSUE THREAD

**Issue by wwolters** [firebase/FirebaseUI-Android#797]
I tried to make an implementation of FirebaseAdapter for ViewPager, but this is not possible due to the RestrictTo annotation on the class. Currently we have to implement the methods without the interface.
It would be nice if the interface were truly public, so all implementations of FirebaseAdapter can be handled generically.

**Comment by samtstern** [comment]
@wwolters is there any reason you're trying to implement `FirebaseAdapter` in particular?  We don't have any methods that accept `FirebaseAdapter` as a param so there shouldn't be any need to use it outside the library.

I think what you really want is to write your own adapter on top of a `FirebaseArray`, which is publicly exposed.

**Comment by TheCraftKid** [comment]
It's recommended to use `FirebaseArray` to implement your own Firebase Database-backed adapters. `FirebaseArray` has all the necessary methods to listen to changes of data at a database location and marshall the data models into whatever is necessary.
For example, I used `FirebaseArray` to make a `ListView` `RemoteViews` adapter because subclassing `ListView` was impossible since it required a special parent class.,

**Comment by wwolters** [comment]
`cleanup` should be called when the adapter is no longer used, so the underlying `FirebaseArray` can be released. We did build an adapter on top of FirebaseArray, it would just be nice to have some sort of marker interface to make it clear `cleanup` should be called, like all official `FirebaseAdapter`s.

**Comment by TheCraftKid** [comment]
We could certainly update the documentation, but AFAIK, there's no platform interface that we could mark `cleanup()` with to show that it needs to be called except `@CallSuper`; however, that only tells the linter that an overriding subclass must call the method, which doesn't apply to composition.

**Comment by SUPERCILEX** [comment]
@samtstern this issue can probably be closed since we don't plan on supporting `FirebaseAdapter` as part of the public api.

@wwolters In FirebaseUI `v2.2.0` we added support for Android Architecture components. I've attached our implementation below, but this should solve your problem by completely removing the need for the client to (un)subscribe from the adapter.
https://github.com/firebase/FirebaseUI-Android/blob/5154ae058f5e276e87a896a78df7aa2802271997/database/src/main/java/com/firebase/ui/database/FirebaseRecyclerAdapter.java#L122-L144
